### PR TITLE
SI-6863 Fix verify error in captured var inited from block w try/catch

### PR DESCRIPTION
**Pull request by JamesIry**
If a captured var was inited from a try/catch we did something
reasonable. But if the var was inited from a block that ended with
a try/catch then we didn't and we were generating faulty byte code.
This fix patches LambdaLift to add the missing case. It also
removes a related, but no longer valid comment from UnCurry. A
test is included.

review @magarciaEPFL

### CODE REVIEW

**Comment by paulp** [comment]
Huge improvement!

File: test/files/run/t6863.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    var coll = {
+      val key = 4
+      try   { null }
+      catch { case _ => null }
+    }
+
+    { () => coll } // the purpose of this is making `coll` an ObjectRef
+
+  }
+}
```
Comment (retronym) [inline test/files/run/t6863.scala]: This case used to crash, too:

``` scala
object Test {
  def main(args: Array[String]) {
    var coll = {
      val key = 4
      if(true) try { null } catch { case _ => null }
      else null
    }

    { () => coll } // the purpose of this is making `coll` an ObjectRef

  }
}
Test.main(null)
```

```
~/code/scala squala sandbox/test.scala
/Users/jason/code/scala/sandbox/test.scala:5: warning: This catches all Throwables, which often has undesirable consequences.
If intentional, use `case _ : Throwable` to clear this warning.
      if(true) try { null } catch { case _ => null }
                                         ^
one warning found
java.lang.VerifyError: (class: Main$$anon$1$Test$, method: main signature: ([Ljava/lang/String;)V) Inconsistent stack height 1 != 3
    at Main$$anon$1.Test$lzycompute(test.scala:1)
```

Comment (JamesIry) [inline test/files/run/t6863.scala]: Yup. That's still a problem. Basically any time a try/catch is in expression position we'll have this problem. Damnit.

Comment (retronym) [inline test/files/run/t6863.scala]: Perhaps we just need to write (or find) `def exprs(t: Tree): List[Tree]` to get all the leaf expressions.

Comment (JamesIry) [inline test/files/run/t6863.scala]: We've got blocks, if/else, and pattern matching. What else?

**Comment by JamesIry** [comment]
Oh, look, I broke it.

File: test/files/run/t6863.scala (hunk 1 above)
Comment (retronym) [inline test/files/run/t6863.scala]: I think you'll need to consider `LabelDef`s, too.

``` scala
squala -Xprint:patmat -e "1 match { case 1 => try 1 finally () }"
...
case <synthetic> val x1: Int = 1;
            case5(){
              if (1.==(x1))
                {
                  <synthetic> val x2: Int = x1;
                  matchEnd4(try {
                    1
                  } finally ())
                }
              else
                case6()
            };
            case6(){
              matchEnd4(throw new MatchError(x1))
            };
            matchEnd4(x: Int){
              x
            }
```

See also #1892

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
Build kitteh is hiding from the vacuum cleaner. For what it's worth, this does compile and pass all tests at https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual/build?delay=0sec

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by adriaanm** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by paulp** [comment]
Wow, this has been a busy little ticket. @JamesIry could I persuade you to squash these four commits? It looks like we have build success and lgtminess so that would do it.

**Comment by JamesIry** [comment]
Yeah, the build kitteh had a spasm on this PR.

Look for a squash coming to a PR near you.

**Comment by JamesIry** [comment]
PLS REBUILD ALL

**Comment by JamesIry** [comment]
The kitteh is happy, @paulp is happy, I'm happy.

**Pull request merged**